### DISCUSSION TITLE
Collecting elements between branches

### DISCUSSION THREAD

**Discussion by oovm** [kevinmehall/rust-peg#394] [Q&A] [answered]
I have a requirement like this in my grammar: if it matches branch a, collect the variables of branch a, if it matches branch b, collect the variables of branch b, and finally merge them together

Roughly similar to the following hypothetical grammar:

```rust
struct Expr {
    a: Option<char>,
    b: Option<char>,
}
peg::parser!( grammar example() for str {
    pub rule select() -> Expr
        = a:a() / b:b() {
        Expr {
            a: Some(a),
            b: Some(b),
        }
    }
    rule a() -> char = "a" { 'a' }
    rule b() -> char = "b" { 'b' }
});
```

I would like to know how to desugar my requirement in `rust-peg`.

In actual use, there are much more complex nested grammar, such as `(a / b) c`

**Comment by kevinmehall** [comment, marked as answer]
I'd probably write that as

```rust
struct Expr {
    a: Option<char>,
    b: Option<char>,
}
peg::parser!( grammar example() for str {
    pub rule select() -> Expr
        = a:a() { Expr { a: Some(a), b: None } }
        / b:b() { Expr { a: None, b: Some(b) } }
   
    rule a() -> char = "a" { 'a' }
    rule b() -> char = "b" { 'b' }
});
```

**Reply by oovm** [reply]
Sorry, I didn't notice that each branch can have a handler.

I'm trying to see if each of my rules can be expanded and rewritten using this pattern.

**Comment by oovm** [comment]
OK, I found that all cases can be solved by two means
- Use the distributive law to expand `/` patterns
- Extract a part as a partial match
  1. If it appears in all branches, it must be T
  2. Otherwise it must be Option<T>, initialized to None in the missing branch
  3. Use the `..` destructuring syntax to merge to the upper layer
